apiserver/facades/controller/caasoperatorprovisioner/provisioner.go

```diff
@@ -57,3 +66,133 @@ func (a *API) WatchApplications() (params.StringsWatchResult, error) {
 	}
 	return params.StringsWatchResult{}, watcher.EnsureErr(watch)
 }
+
+// UpdateApplicationsUnits updates the Juju data model to reflect the given
+// units of the specified application.
+func (a *API) UpdateApplicationsUnits(args params.UpdateApplicationUnitArgs) (params.ErrorResults, error) {
+	result := params.ErrorResults{
+		Results: make([]params.ErrorResult, len(args.Args)),
+	}
+	if len(args.Args) == 0 {
+		return result, nil
+	}
+	for i, appUpdate := range args.Args {
+		appTag, err := names.ParseApplicationTag(appUpdate.ApplicationTag)
+		if err != nil {
+			result.Results[i].Error = common.ServerError(err)
+			continue
+		}
+		app, err := a.state.Application(appTag.Id())
+		if err != nil {
+			result.Results[i].Error = common.ServerError(err)
+			continue
+		}
+		if err := a.updateUnitsFromCloud(app, appUpdate.Units); err != nil {
+			result.Results[i].Error = common.ServerError(err)
+		}
+	}
+	return result, nil
+}
+
+// UpdateUnitsFromCloud takes a slice of unit information provided by an external
+// source (typically a cloud update event) and merges that with the existing unit
+// data model in state.
+func (a *API) updateUnitsFromCloud(app Application, units []params.ApplicationUnitParams) error {
+	// Set up the initial data structures.
+	existingStateUnits, err := app.AllUnits()
+	if err != nil {
+		return errors.Trace(err)
+	}
+
+	var (
+		// unassociatedUnits are those which do noy yet
+		// have a UUID assigned.
+		unassociatedUnits []Unit
+
+		// removedUnits are those that exist in state but are not
+		// represented in the unit params passed in.
+		removedUnits []Unit
+
+		addedCloudUnits []params.ApplicationUnitParams
+	)
+	stateUnitsById := make(map[string]Unit)
+	cloudUnitIds := make(set.Strings)
+
+	existingCloudUnitsById := make(map[string]params.ApplicationUnitParams)
+
+	// Record all unit uuids known to exist in the cloud.
+	for _, u := range units {
+		cloudUnitIds.Add(u.Id)
+	}
+
+	// Loop over any existing state units and record those which do not yet have
+	// provider ids, and those which have been removed or updated.
+	for _, u := range existingStateUnits {
+		if u.ProviderId() == "" {
+			logger.Debugf("unit %q is unallocated", u.Name())
+			unassociatedUnits = append(unassociatedUnits, u)
+		} else if u.Life() == state.Alive {
+			if cloudUnitIds.Contains(u.ProviderId()) {
+				logger.Debugf("unit %q (%v) has changed in the cloud", u.Name(), u.ProviderId())
+				stateUnitsById[u.ProviderId()] = u
+			} else {
+				logger.Debugf("unit %q (%v) has been removed from the cloud", u.Name(), u.ProviderId())
+				removedUnits = append(removedUnits, u)
+			}
+		}
+	}
+	for _, u := range units {
+		if _, ok := stateUnitsById[u.Id]; ok {
+			existingCloudUnitsById[u.Id] = u
+		} else {
+			addedCloudUnits = append(addedCloudUnits, u)
```

Comment:
what if the unit already exists in state, but is Dying? I don't think we would want to add it then?

Reply:
There was a check for life in the "for _, u := range existingStateUnits" loop but I've moved it so all not alive units are filtered out.